src/System.Text.Primitives/System/Text/TextEncoding.cs

```diff
@@ -121,6 +123,156 @@ public override bool TryEncodeFromUtf16(ReadOnlySpan<char> utf16, Span<byte> buf
             return true;
         }
 
+        public override bool TryEncodeFromUnicode(ReadOnlySpan<UnicodeCodePoint> codePoints, Span<byte> buffer, out int bytesWritten)
+        {
+            var avaliableBytes = buffer.Length;
+            int bytesWrittenForCodePoint = 0;
+            bytesWritten = 0;
+
+            for (int i = 0; i < codePoints.Length; i++)
+            {
+                if (!TryEncodeFromUnicode(codePoints[i], buffer.Slice(bytesWritten), out bytesWrittenForCodePoint))
+                {
+                    bytesWritten = 0;
+                    return false;
+                }
+                bytesWritten += bytesWrittenForCodePoint;
+            }
+
+            return true;
+        }
+
+        public override bool TryEncodeFromUnicode(UnicodeCodePoint codePoint, Span<byte> buffer, out int bytesWritten)
+        {
+            var avaliableBytes = buffer.Length;
+            bytesWritten = 0;
+
+            var codepoint = (uint)codePoint;
+            if (codepoint <= 0x7f) // this if block just optimizes for ascii
+            {
+                if (bytesWritten + 1 > avaliableBytes)
+                {
+                    bytesWritten = 0;
+                    return false;
+                }
+                buffer[bytesWritten++] = (byte)codepoint;
+            }
+            else
+            {
+                Utf8EncodedCodePoint encoded;
+                encoded = new Utf8EncodedCodePoint(codePoint);
+
+                if (bytesWritten + encoded.Length > avaliableBytes)
+                {
+                    bytesWritten = 0;
+                    return false;
+                }
+
+                buffer[bytesWritten] = encoded.Byte0;
+                if (encoded.Length > 1)
+                {
+                    buffer[bytesWritten + 1] = encoded.Byte1;
+
+                    if (encoded.Length > 2)
+                    {
+                        buffer[bytesWritten + 2] = encoded.Byte2;
+
+                        if (encoded.Length > 3)
+                        {
+                            buffer[bytesWritten + 3] = encoded.Byte3;
+                        }
+                    }
+                }
+
+                bytesWritten += encoded.Length;
+            }
+            return true;
+        }
+
+        public override bool TryDecodeToUnicode(Span<byte> encoded, Span<UnicodeCodePoint> decoded, out int bytesWritten)
```

Comment:
The existing Encoder/Decoder.Convert is pretty close to what you need (without the `completed` out argument because of your encoders are state free):

https://github.com/dotnet/corefx/blob/master/src/System.Runtime/ref/System.Runtime.cs#L7046